### ISSUE TITLE
Hardware FRAM IC revision/upgrade to 8Mbit

### ISSUE THREAD

**Issue by Firemission1343** [machdyne/blaustahl#13]
First off very cool project! Found your project from LTT video and wanted to look more into it. 

Currently from what I can tell, the FRAM IC used is the MB85RS64 (blaustahl_v0.pdf says MB85RS16) which is at 64Kbit ([Datasheet](https://www.lcsc.com/datasheet/lcsc_datasheet_2112091730_FUJITSU-MB85RS64PNF-G-JNERE1_C8741.pdf)).
The largest other IC in the same 8-SOIC package is a 8Mbit CY15B108QN-40SXI from Infineon Technologies ([Datasheet](https://www.infineon.com/dgdl/Infineon-CY15B108QN_CY15V108QN_Excelon(TM)_LP_8-Mbit_(1024K_X_8)_Serial_(SPI)_F-RAM-DataSheet-v11_00-EN.pdf?fileId=8ac78c8c7d0d8da4017d0ee7134b6ff4)).

It would be quite an upgrade from the 64Kbit size limitation, to a 8Mbit at 125x increase in memory at about only 20x in cost (ignoring other costs from board changes)
There are some important differences like clock frequency, supply voltage, and Pin 7 but would it be feasible to create a V2 using a larger FRAM ic? 
Personally, the 8KB is just a tad bit too limited for any uses cases I would want to use.

**Comment by Supermath101** [comment]
With FRAM chips that are that big on the market, such as the RAMXEED MB85RQ8MX ([Datasheet](https://www.ramxeed.com/assets/images/products/datasheet/FeRAM/s1/MB85RQ8MX-DS1v1-E.pdf)), you could likely ditch the on-board QSPI flash chip altogether, and only have the rp2040, FRAM, and other miscellaneous components.

**Comment by machdyne** [comment]
@Firemission1343 I am planning a higher capacity version, possibly with a different name and different firmware. A filesystem starts to make sense at some point. Blaustahl does use MB85RS64, the MB85RS16 was used in [Stahl](https://github.com/machdyne/stahl).

@Supermath101 I have been experimenting with all-FRAM systems as part of the [Zwölf MCU](https://machdyne.com/2024/11/15/zwolf-mcu-platform/) project. I need to test RP2040 boot compatibility with the higher capacity FRAM, but even if it doesn't work out there are FRAM-based MCUs that could make an all-FRAM device possible.

Thanks.

**Comment by machdyne** [comment]
An update for anyone interested in higher capacity FRAM storage ...

I've posted some details about a higher capacity version of Blaustahl:

https://machdyne.com/product/kaltstahl-storage-device/

Depending on demand I may hand-assemble a limited number of those.

I'm also working on a long-term storage array called Ebenstahl:

https://machdyne.com/product/ebenstahl-storage-device/

**Comment by akrabu** [comment]
Getting stuck on the CAPCHA trying to reserve one!

![Image](https://github.com/user-attachments/assets/e30939a6-5f84-4f2d-a35c-09b9c5561aae)

**Comment by tailz303** [comment]
Just curious if at all possible is there some sort of hexedit or bin editing that could be done to allow bigger chips if needed? I have ordered a MB85RS4MTYPF-G-BCERE1 from mouser and am waiting for it to arrive. It is a seemingly compatible spi 4mbit sop8 chip. I looked at the firmware and tried to run strings, use uf2conv.py and some other noobile things I could think of on it to see if I could find some easy to modify logic that limits the max size. Currently its just for fun and since I have a ch341a_spi and matching sop8 clip (mainly for bios flashing) I can always undo something if broken. Basically all that to say a 500,000 ish character serial text editor f-ram thingy would be funny and cool and I want to do it but dont know if the firmware will get mad if I swap the chip.

**Comment by machdyne** [comment]
Rebuilding the firmware is probably much easier, you would just need to define FRAM_BIG and increase FRAM_PAGES in blaustahl.h.

FRAM_BIG uses 24-bit addresses instead of 16-bit addresses when talking to the FRAM, which is necessary on the larger FRAMs. Because of that the default firmware won't work with the larger chips.

Keep in mind the Blaustahl is virtually unbrickable thanks to the RP2040 bootrom, you should be able to use the boot select button to access the UF2 bootloader even if the flash becomes corrupt.

As far as the MB85RS4MTYPF-G-BCERE1 is concerned, I believe it is 208mil wide which won't probably won't quite fit onto the footprint but you may be able to bend the pins and get it working.

[Kaltstahl](https://machdyne.com/product/kaltstahl-storage-device/) is using MB85RS2MTAPNF-G-BDERE1 which is the largest FRAM I was able to find in a 150mil wide SOP-8 package.

If you have trouble building the firmware let me know and I will make available some binaries for larger FRAM sizes.

**Comment by tailz303** [comment]
I might have done something wrong. I first tested with big_fram uncommented and 16 pages, this cascaded the existing text like 40 times down the screen so I put it back to 4, same thing, then disabled big_fram all together and it seemed to fix the issue. Might I have done something wrong or does fram_big only support actually big frams? I will get more info when I get the 4mbit flash. I will try what you said and bend the pins to fit since it is 208mil

**Comment by machdyne** [comment]
FRAM_BIG will only work correctly with the bigger FRAMs, because it sends a 24-bit address and the smaller (8KB) FRAM expects at 16-bit address. Please let us know how the upgrade goes. Thanks.

**Comment by tailz303** [comment]
After much waiting for UPS here's the results with pictures. It worked!

![Image](https://github.com/user-attachments/assets/453a26d3-1894-4b32-939f-441a4a246135)
![Image](https://github.com/user-attachments/assets/d2721437-a3cb-4336-b90b-db8a08cdc705)
![Image](https://github.com/user-attachments/assets/ec3bc923-dff8-4438-b8c3-8adf2fa7b634)
![Image](https://github.com/user-attachments/assets/62867a8b-1c45-4971-9472-d5436de6dacb)
![Image](https://github.com/user-attachments/assets/f47f7342-919b-4524-8e72-5b6e59278724)

I tested and data holds integrity at page 1, 108, and 250 whereas 8 pages on the 8KB chip would repeat every 4th page. This verifies that it's communicating correctly

**Comment by machdyne** [comment]
That's great, glad to see that the wider chip fits.

**Comment by rcmcdonald91** [comment]
@tailz303 do you have a link for these bigger chips? 

also #23

**Comment by tailz303** [comment]
> That's great, glad to see that the wider chip fits.

Yeah it was pretty neat, I was able to take the "sprawl" of the legs and fold them under the chip essentially making a worse less flat wson8 chip but it fit well on the pad layout and aside from my imperfect bending soldered on well! I would have fixed it but didn't want to keep heating up the board. 





> [@tailz303](https://github.com/tailz303) do you have a link for these bigger chips?

Yes! here's the one I used: 'https://www.mouser.com/ProductDetail/Ramxeed/MB85RS4MTYPF-G-BCERE1?qs=HBWAp0VN4Rjdon2wvh8hTQ%3D%3D&countryCode=US&currencyCode=USD'

Edit: here's the pre-compiled firmware with fram_big defined and 250 pages configured (probably is supposed to be 256 but whatever) in case you get the same one and don't want to compile it again, if you get an 8mbit fram I imagine it'll need to be compiled with 500 pages (maybe 512).
'https://files.happyfile.net/uploads/blaustahl_10.88.65.48_happyfile.net_cbb828f4.uf2' non cdonly version
'https://files.happyfile.net/uploads/blaustahl_cdconly_10.88.65.48_happyfile.net_d9cb3cd0.uf2' cdonly version (seems like the right one for most things based on the docs)

Edit 2: Link to mouser search with "chips that should be compatible" but verify before you buy.
'https://www.mouser.com/c/semiconductors/memory-ics/f-ram/?memory%20size=1%20Mbit~~8%20Mbit&package%20%2F%20case=SOIC-8%7C~SOP-8&rp=semiconductors%2Fmemory-ics%2Ff-ram%7C~Memory%20Size'

**Comment by rcmcdonald91** [comment]
@tailz303 thanks!!
